clippy_lints/src/let_arr_const.rs

```diff
@@ -0,0 +1,116 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::implements_trait;
+use clippy_utils::visitors::is_const_evaluatable;
+use hir::{BindingMode, ExprKind, PatKind, Stmt, StmtKind};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for defining of read-only arrays on stack.
+    ///
+    /// ### Why is this bad?
+    /// A read-only array should be declared as a `static` item or used tricks
+    /// to made it into `.rodata` section of the compiled file.
+    ///
+    /// ### Known problems
+    /// `let array` puts array on the stack which might make the generated binary file
+    /// bigger and slower.
```

Comment:
This should go into the "Why is this bad?" section. "Known problems" is for problems with this lint (ie false positives or false negatives).

But also, this should explain more concretely why this makes the binary slower or faster.

For example, what initially confused me from reading this was that `*&<array literal>` looked like it would still have to copy it onto the stack, so you'd have the array end up on the stack either way (semantically, ignoring opts).

The way I understand it is that the address-of operator makes rustc static-promote the array literal, which means that it can avoid emitting individual movs or memsets that would be required to initialize the array. Then dereferencing it can emit a memcpy without having to mutate the copy (unlike before where it would need `store`s to initialize the array), which gives LLVM an easier time to reason about it and optimize that copy away if you only e.g. directly index into it?